File: Medium/2025-12-26-2483-Minimum-Penalty-for-a-Shop/solution.js
var bestClosingTime = function(customers) {
    let penalty = 0;
    let minPenalty = Infinity;
    let bestHour = 0;

    for (let i = 0; i < customers.length; i++) {
        if (customers[i] === 'N') {
            penalty++;
        }
    }

    for (let j = 0; j <= customers.length; j++) {
        if (penalty < minPenalty) {
            minPenalty = penalty;
            bestHour = j;
        }
        if (j < customers.length) {
            if (customers[j] === 'Y') {
                penalty++;
            } else {
                penalty--;
            }
        }
    }

    return bestHour;
};

File: Medium/2025-12-26-2483-Minimum-Penalty-for-a-Shop/solution.py
class Solution:
    def bestClosingTime(self, customers: str) -> int:
        n = len(customers)
        penalty = [0] * (n + 1)
        
        for i in range(n):
            penalty[i + 1] = penalty[i] + (customers[i] == 'N')
        
        min_penalty = float('inf')
        best_time = 0
        
        for j in range(n + 1):
            current_penalty = penalty[j] + (n - j - (penalty[n] - penalty[j]))
            if current_penalty < min_penalty:
                min_penalty = current_penalty
                best_time = j
        
        return best_time